pkg/distributor/distributor.go

```diff
@@ -1054,13 +1056,11 @@ func (d *Distributor) limitsMiddleware(next push.Func) push.Func {
 		if !d.requestRateLimiter.AllowN(now, userID, 1) {
 			d.discardedRequestsRateLimited.WithLabelValues(userID).Add(1)
 
-			// Return a 429 or a 529 here depending on configuration to tell the client it is going too fast.
-			// Client may discard the data or slow down and re-send.
-			// Prometheus v2.26 added a remote-write option 'retry_on_http_429'.
-			if d.limits.ServiceOverloadStatusCodeOnRateLimitEnabled(userID) {
-				return httpgrpc.Errorf(statusServiceOverload, validation.NewRequestRateLimitedError(d.limits.RequestRate(userID), d.limits.RequestBurstSize(userID)).Error())
-			}
-			return httpgrpc.Errorf(http.StatusTooManyRequests, validation.NewRequestRateLimitedError(d.limits.RequestRate(userID), d.limits.RequestBurstSize(userID)).Error())
+			return distributorerror.NewRequestRateError(
+				d.limits.RequestRate(userID),
+				d.limits.RequestBurstSize(userID),
+				d.limits.ServiceOverloadStatusCodeOnRateLimitEnabled(userID),
```

Comment:
Can you please create an issue to move this decision to the API/Handler level? The error here should know nothing about the status code we want to return on one of the transport layers.

Reply:
I have fixed this in the present PR.